core/transaction-pool/graph/src/pool.rs

```diff
@@ -286,200 +210,236 @@ impl<B: ChainApi> Pool<B> {
-		self.remove_invalid(&to_remove);
-		self.remove_invalid(&futures_to_remove);
-		// clear banned transactions timeouts
-		self.rotator.clear_timeouts(&now);
-
-		Ok(())
-	}
-
-	/// Create a new transaction pool.
-	pub fn new(options: Options, api: B) -> Self {
-		Pool {
-			api,
-			options,
-			listener: Default::default(),
-			pool: Default::default(),
-			import_notification_sinks: Default::default(),
-			rotator: Default::default(),
-		}
+	) -> impl Future<Output=Result<(), B::Error>> {
+		self.validated_pool.prune_tags(at, tags, known_imported_hashes)
 	}
 
 	/// Return an event stream of transactions imported to the pool.
 	pub fn import_notification_stream(&self) -> EventStream {
-		let (sink, stream) = mpsc::unbounded();
-		self.import_notification_sinks.lock().push(sink);
-		stream
+		self.validated_pool.0.import_notification_stream()
 	}
 
 	/// Invoked when extrinsics are broadcasted.
 	pub fn on_broadcasted(&self, propagated: HashMap<ExHash<B>, Vec<String>>) {
-		let mut listener = self.listener.write();
-		for (hash, peers) in propagated.into_iter() {
-			listener.broadcasted(&hash, peers);
-		}
+		self.validated_pool.0.on_broadcasted(propagated)
 	}
 
 	/// Remove from the pool.
 	pub fn remove_invalid(&self, hashes: &[ExHash<B>]) -> Vec<TransactionFor<B>> {
-		// temporarily ban invalid transactions
-		debug!(target: "txpool", "Banning invalid transactions: {:?}", hashes);
-		self.rotator.ban(&time::Instant::now(), hashes.iter().cloned());
-
-		let invalid = self.pool.write().remove_invalid(hashes);
-
-		let mut listener = self.listener.write();
-		for tx in &invalid {
-			listener.invalid(&tx.hash);
-		}
-
-		invalid
+		self.validated_pool.0.remove_invalid(hashes)
 	}
 
 	/// Get an iterator for ready transactions ordered by priority
 	pub fn ready(&self) -> impl Iterator<Item=TransactionFor<B>> {
-		self.pool.read().ready()
+		self.validated_pool.0.ready()
 	}
 
 	/// Returns pool status.
 	pub fn status(&self) -> base::Status {
-		self.pool.read().status()
+		self.validated_pool.0.status()
 	}
 
 	/// Returns transaction hash
 	pub fn hash_of(&self, xt: &ExtrinsicFor<B>) -> ExHash<B> {
-		self.api.hash_and_length(xt).0
+		self.validated_pool.0.api().hash_and_length(xt).0
 	}
 }
 
-fn fire_events<H, H2, Ex>(
-	listener: &mut Listener<H, H2>,
-	imported: &base::Imported<H, Ex>,
-) where
-	H: hash::Hash + Eq + traits::Member + Serialize,
-	H2: Clone,
-{
-	match *imported {
-		base::Imported::Ready { ref promoted, ref failed, ref removed, ref hash } => {
-			listener.ready(hash, None);
-			for f in failed {
-				listener.invalid(f);
-			}
-			for r in removed {
-				listener.dropped(&r.hash, Some(hash));
-			}
-			for p in promoted {
-				listener.ready(p, None);
-			}
-		},
-		base::Imported::Future { ref hash } => {
-			listener.future(hash)
-		},
+impl<B: ChainApi> Clone for SharedValidatedPool<B> {
+	fn clone(&self) -> Self {
+		Self(self.0.clone())
+	}
+}
+
+impl<B: ChainApi> SharedValidatedPool<B> {
```

Comment:
Why do we implement the methods for `SharedValidatedPool` and not `Pool`?

Reply:
That's an artefact of previous implementation - merged methods of both in single `Pool` struct.